Dockerfile

```diff
@@ -23,19 +23,30 @@ RUN mkdir /src/build && \
     cmake --build . && \
     cmake --build . --target install
 
-FROM debian:bullseye-slim
+FROM debian:bookworm-slim
 
-RUN apt-get update && \
-    apt-get install -y \
-        binutils \
+COPY --from=builder /usr/local/bin/kcov* /usr/local/bin/
+COPY --from=builder /usr/local/share/doc/kcov /usr/local/share/doc/kcov
+
+RUN set -eux; \
+    apt-get update; \
+    apt-get install -y --no-install-recommends --no-install-suggests \
         libcurl4 \
         libdw1 \
         zlib1g \
-        && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/*
-
-COPY --from=builder /usr/local/bin/kcov* /usr/local/bin/
-COPY --from=builder /usr/local/share/doc/kcov /usr/local/share/doc/kcov
+    ; \
+    apt-get clean; \
+    rm -rf /var/lib/apt/lists/*; \
+    # Write a test script
+    echo '#!/usr/bin/env bash\nif [[ true ]]; then\necho "Hello, kcov!"\nfi' > /tmp/test-executable.sh; \
```

Comment:
This script is also available as `.github/workflows/test-executable.sh`, so might be possible to simply copy from there.

However, I think a better test (although more fickle) would be to test against a compiled binary, since the shell script code isn't architecture-specific. That would require the docker builder to run with `--security-opt seccomp=unconfined` though, since ptrace normally isn't available via docker.

Reply:
> This script is also available as .github/workflows/test-executable.sh, so might be possible to simply copy from there.

Yes but if I do that it will be part of the final image, not very clean..

> However, I think a better test (although more fickle) would be to test against a compiled binary, since the shell script code isn't architecture-specific. That would require the docker builder to run with --security-opt seccomp=unconfined though, since ptrace normally isn't available via docker.

Do you have a simple binary in mind ?

Reply:
A simple

```
int main(int argc, const char *argv[])
{
   printf("Hello world!\n");
   if (argc > 1000)
   {
      printf("Unacceptable!\n");
   } 
   return 0;
}
```

will do just fine. It's basically how the automated tests are done, where the cobertura file is parsed to verify that e.g., the "Hello world" line is executed, but that the "Unacceptable!" line isn't. There's a helper script, `tests/tools/parse_cobertura` which does the lookup, given a file and line pair.

It doesn't work for the non-x86 github action runners since these run via qemu-user, where ptrace isn't possible to use. They instead just check that a basic shell script works, like in your Dockerfile.

For docker images, one instead needs the `--security-opt` mentioned above.